qir/qat/PreTransformTrimming/PreTransformTrimmingPass.cpp

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "PreTransformTrimming/PreTransformTrimmingPass.hpp"
+
+#include "Llvm/Llvm.hpp"
+
+#include <fstream>
+#include <iostream>
+
+namespace microsoft
+{
+namespace quantum
+{
+
+    PreTransformTrimmingPass::PreTransformTrimmingPass(
+        PreTransformTrimmingPassConfiguration const& cfg,
+        ILoggerPtr const&                            logger)
+      : config_{cfg}
+      , logger_{logger}
+    {
+    }
+
+    llvm::PreservedAnalyses PreTransformTrimmingPass::run(llvm::Module& module, llvm::ModuleAnalysisManager& /*mam*/)
+    {
+
+        std::vector<llvm::Function*> deletables;
+        auto                         entry_point_attr = config_.entryPointAttr();
+        for (auto& function : module)
+        {
+
+            if (function.isDeclaration())
+            {
+                continue;
+            }
+
+            if (function.hasFnAttribute(entry_point_attr))
+            {
+                continue;
+            }
+
+            deletables.push_back(&function);
+        }
+
+        for (auto& x : deletables)
+        {
+            if (x->use_empty())
+            {
+                x->eraseFromParent();
```

Comment:
The code deletes all functions which are not referenced other places in the code. That is, it is a simplified dead code elimination pass which _only_ trims unused functions. I am not sure how this is contradictory to later having an inline pass. Two functions will indeed still exist in the code after running this pass if one function is calling the other. Maybe I am missing something here? Regardless, note that you can always turn this functionality off by just using the `--disable-` option. Default is on though as trimming dead code prior to performing transformation has several upsides.

Reply:
Ah, that's what I was misunderstanding: I didn't realize that `use_empty` meant it would only delete a function that was never called. That completely resolves my concern, thanks for pointing that out.